app/src/androidTest/java/com/github/se/polyfit/data/api/OpenFoodFactsApiCallerTest.kt

```diff
@@ -0,0 +1,162 @@
+package com.github.se.polyfit.data.api
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.github.se.polyfit.model.nutritionalInformation.MeasurementUnit
+import com.github.se.polyfit.model.nutritionalInformation.Nutrient
+import com.github.se.polyfit.model.nutritionalInformation.NutritionalInformation
+import okhttp3.mockwebserver.Dispatcher
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+import org.json.JSONObject
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class OpenFoodFactsApiCallerTest {
+  @get:Rule val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+  private lateinit var mockWebSever: MockWebServer
+  private lateinit var foodFactsApi: OpenFoodFactsApi
+
+  private val nutellaCode: String = "3017624010701"
+
+  private val dispatcher =
+      object : Dispatcher() {
+        override fun dispatch(request: RecordedRequest): MockResponse {
+          return when (request.path) {
+            "/$nutellaCode?fields=product_name,nutriments,quantity" ->
+                MockResponse().setResponseCode(200).setBody(nutellaResult.toString())
```

Comment:
Good job testing with both a normal and faulty dispatcher. Something else you could test is how your code reacts to a body being empty or null. This happens with the real API if the QR code is invalid or something similar.

Reply:
oh right forgot this case